### ISSUE TITLE
Run JSDoc in the check step

### ISSUE THREAD

**Issue by NoraCodes** [OpenEnergyDashboard/OED#396]
The check step should include JSDoc and/or TSDoc to prevent people from writing broken documentation.

**Comment by huss** [comment]
After this is added, we need to fix any broken JSDoc.

**Comment by huss** [comment]
The first step will be to get the documentation checker to run as part of the code checks done (see package.json check scripts for what is currently done). This will undoubtedly lead to lots of code with inadequate documentation. This can be addressed now or made into an issue for the future. Note if parts of OED does not pass this added test then we want it to run as part of the checks but not have it cause the build to fail at this time. See .github/workflows/build.yml for information on the automatic tests run on GitHub.

**Comment by huss** [comment]
This might be enabled or partly enabled. Need to check.

**Comment by rltsystems** [comment]
@bryhernandez and I have been taking a look through the package.json. 
We're currently unable to find any scripts that are being included in package.json and the build.yml file that are checking (validation) for JSDoc comments in the .JS / .JSX files. 
@huss Please advise if this is what the expectation for this issue is. 

We are finding that the wording of the issue is a bit ambiguous, thus we spent a good amount of time backtracking, as this is our initial exposure to these processes.

We have identified params that will be required for grep statements in the script(s) but are researching how to best test/implement them currently. 

**JSDOC rules consist of params like:**  

- Multiline comments should start with /** and end with */, and in between should start with *
- Tags should only consist of the following tags:
**Types**
[@type](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#type)
[@param](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns) (or [@arg](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns) or [@argument](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns))
[@returns](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns) (or [@return](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns))
[@typedef](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#typedef-callback-and-param)
[@callback](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#typedef-callback-and-param)
[@template](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template)[](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#classes)
**Classes**
[Property Modifiers](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#property-modifiers) @public, @private, @protected, @readonly
[@override](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#override)
[@extends](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#extends) (or [@augments](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#extends))
[@implements](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#implements)
[@class](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#constructor) (or [@constructor](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#constructor))
[@this](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#this)[](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#documentation)
**Documentation**
[@deprecated](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated)
[@see](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#see)
[@link](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#link)[](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#other)
**Other**
[@enum](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#enum)
[@author](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#author)

**Comment by bryhernandez** [comment]
> @bryhernandez and I have been taking a look through the package.json. We're currently unable to find any scripts that are being included in package.json and the build.yml file that are checking (validation) for JSDoc comments in the .JS / .JSX files. @huss Please advise if this is what the expectation for this issue is.
> 
> We are finding that the wording of the issue is a bit ambiguous, thus we spent a good amount of time backtracking, as this is our initial exposure to these processes.
> 
> We have identified params that will be required for grep statements in the script(s) but are researching how to best test/implement them currently.
> 
> **JSDOC rules consist of params like:**
> 
> * Multiline comments should start with /** and end with */, and in between should start with *
> * Tags should only consist of the following tags:
>   **Types**
>   [@type](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#type)
>   [@param](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns) (or [@arg](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns) or [@argument](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns))
>   [@returns](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns) (or [@return](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns))
>   [@typedef](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#typedef-callback-and-param)
>   [@callback](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#typedef-callback-and-param)
>   [@template](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template)
>   **Classes**
>   [Property Modifiers](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#property-modifiers) @public, @Private, @Protected, @readonly
>   [@override](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#override)
>   [@extends](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#extends) (or [@augments](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#extends))
>   [@implements](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#implements)
>   [@class](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#constructor) (or [@constructor](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#constructor))
>   [@this](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#this)
>   **Documentation**
>   [@deprecated](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated)
>   [@see](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#see)
>   [@link](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#link)
>   **Other**
>   [@enum](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#enum)
>   [@author](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#author)

Hey, @rltsystems I found this [stack overflow article](https://stackoverflow.com/questions/67036753/grep-regex-match-only-if-a-line-doesnt-start-by) that can help us on this issue. 

We can definitely use this as a building block to create a check script.

**Comment by rltsystems** [comment]
More resources for @bryhernandez and I on code documentation guidelines on the project:
https://openenergydashboard.github.io/developer/codeDocumentation.html

**Comment by rltsystems** [comment]
We found an NPM Packaged called eslint-plugin-jsdoc that may offer functionality to provide JSdoc documentation checking. It seems to work alongside ESLint (https://www.npmjs.com/package/eslint) so we figure that we won't have to manually write script files w/ grep/awk etc. in order to check JSDoc.

https://www.npmjs.com/package/eslint-plugin-jsdoc 

The next steps are to test this package, read through the documentation to better understand its' features and usage, and determine feasibility for integrating into the project.

**Comment by rltsystems** [comment]
@huss Please advise. 

We were initially confused about the layout since there was a lack of documentation however we realized that because there is another issue #463 that includes migrating TSLint to ESLint (Node pkg), so we were initially checking to see if ESLint and eslint-plugin-jsdoc were setup, however it was unrelated to the current issue. We believe this issue just involved checking the current TSLint was running in the check step. 

@bryhernandez  and I have confirmed that it is currently running via these three files --> tslint.json --> Listed as "check: lint" in the package.json file alongside the proper file type (i.e. .js .tsx . ts .jsx) --> check:lint is run in the check step on the build.yml file in .github/workflows directory.

![image](https://user-images.githubusercontent.com/10646650/159782688-9311ebcb-abad-4909-8626-1d59e3d171f8.png)
![image](https://user-images.githubusercontent.com/10646650/159782815-f40c9cbf-6001-4a04-aa0b-6c81406cac2c.png)

**Comment by huss** [comment]
Thanks for the info. First, someone is working on upgrading to ESLint right now as part of package upgrades. The hope is that will happen in the near future. Second, we are running TSLint but does that do any JSDoc/TSDoc checks? If so, we are okay but I have not seen any such message so it is possible it is not doing those checks. As you have found, ESLint may make it easy to do the Doc checks and knowing how to do that as we migrate to ESLint would be helpful.

**Comment by rltsystems** [comment]
@huss 
We looked into it a bit more and found that tslint.json enforces javascript rules currently, and saw that jsdoc is not included as a rule. We think that using the jsdoc-format rule should allow us this functionality so we'll test it locally first.

Supporting documentation available here: https://palantir.github.io/tslint/rules/jsdoc-format/

**Comment by rltsystems** [comment]
@huss TSLint is now deprecated as of #463 , thus we have moved on to the implementation of the JSDOC check via a node package that utilizes ESLint (eslint-plugin-jsdoc)

@bryhernandez and I saw that @jina2k had a successful merge/resolution of issue #463 so we forked the latest version of the repo, created our own branch based on the development branch (JSDoc_Check), and then installed the eslint-plugin-jsdoc node package successfully.
- [**eslint-plugin-jsdoc Node Package Documentation**](https://www.npmjs.com/package/eslint-plugin-jsdoc#user-content-eslint-plugin-jsdoc-installation)

- We will be submitting a pull request from our branch (JSDoc_Check) shortly, after commenting out errors to make sure the build is not blocked by the errors generated.

Some sample output after running **npm run check**
![JSDoc_check Example Output](https://user-images.githubusercontent.com/10646650/162525238-17d1e979-7152-48e6-98a4-bd216a44d1ac.png)
See JSDoc_check-output.txt for an example of the output.
[JSDoc_check-output.txt](https://github.com/OpenEnergyDashboard/OED/files/8455049/JSDoc_check-output.txt)
